### PR TITLE
Sub processor push down part1

### PR DESCRIPTION
**Pull request by robstryker**
## What it does

Separate core and ui for GetterSetterCorrectionSubProcessor and VarargsWarningsSubProcessor. 

## How to test

Any UI entrypoint that would access the various code completions and quick fixes should be able to trigger this code. 

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java [hunk 1]
```diff
@@ -0,0 +1,363 @@
+/*******************************************************************************
+ * Copyright (c) 2007, 2016 IBM Corporation and others.
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java:2]: Fix dates on new files - global comment (should be just 2024)

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/JavadocTagsBaseSubProcessor.java [hunk 2]
```diff
@@ -267,14 +278,18 @@ public void addMissingJavadocCommentProposals(IInvocationContextCore context, IP
 				}
 				if (comment != null) {
 					String label= CorrectionMessages.JavadocTagsSubProcessor_addjavadoc_field_description;
-					proposals.add(addJavadocCommentProposal(label, cu, IProposalRelevance.ADD_JAVADOC_FIELD, declaration.getStartPosition(), comment));
+					T p = addJavadocCommentProposal(label, cu, IProposalRelevance.ADD_JAVADOC_FIELD, declaration.getStartPosition(), comment);
+					if( p != null )
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/JavadocTagsBaseSubProcessor.java:282]: global comment - stick to existing if formatting.  Have if (expression) rather than jamming open parenthesis against if and adding spaces around expression

**Review by jjohnstn — CHANGES_REQUESTED** [review]
Minor nit comments.  Copyright year and some formatting changes

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+		public static Change getChange(IField field) throws CoreException {
+			Refactoring refactoring= getChangeRefactoring(field);
+			Change createdChange= refactoring.createChange(new NullProgressMonitor());
+			return createdChange;
+		}
+
+		public static TextFileChange getChange(IField field, IFile file) throws CoreException {
+			Change createdChange= getChange(field);
+			if (createdChange instanceof CompositeChange) {
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java:123]: Use pattern matching here and save the cast on the next line.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+				for (Change curr : ((CompositeChange) createdChange).getChildren()) {
+					if (curr instanceof TextFileChange && (file == null || ((TextFileChange) curr).getFile().equals(file))) {
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java:125]: Use pattern matching here and save the cast on this and the next line.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+		}
+
+
+		public static String getDescription(IField field) {
+			return Messages.format(CorrectionMessages.GetterSetterCorrectionSubProcessor_creategetterunsingencapsulatefield_description, BasicElementLabels.getJavaElementName(field.getElementName()));
+		}
+
+		/*
+		 * @see org.eclipse.jface.text.contentassist.ICompletionProposalExtension5#getAdditionalProposalInfo(org.eclipse.core.runtime.IProgressMonitor)
+		 * @since 3.5
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java:140]: Please remove the whole comment here:
* Current version of the bundle is 1.21 thus the since is totally wrong. 
* The see tag is useless either as ICompletionProposalExtension5 is not visible in this bundle.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+		}
+		return addGetterSetterProposals(context, coveringNode, resultingCollections, IProposalRelevance.GETTER_SETTER_QUICK_ASSIST);
+	}
+
+	public void addGetterSetterProposals(IInvocationContextCore context, IProblemLocationCore location, Collection<T> proposals, int relevance) {
+		addGetterSetterProposals(context, location.getCoveringNode(context.getASTRoot()), proposals, relevance);
+	}
+
+	protected boolean addGetterSetterProposals(IInvocationContextCore context, ASTNode coveringNode, Collection<T> proposals, int relevance) {
+		if (!(coveringNode instanceof SimpleName)) {
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java:177]: Use pattern matching here and save the line doing the cast.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+			return false;
+		}
+		SimpleName sn= (SimpleName) coveringNode;
+
+		IBinding binding= sn.resolveBinding();
+		if (!(binding instanceof IVariableBinding))
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java:183]: Use pattern matching here and save the line doing the cast.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+		IMethodBinding method= findGetter(context);
+		if (method != null) {
+			Expression mi= createMethodInvocation(context, method, null);
+			context.astRewrite.replace(context.accessNode, mi, null);
+
+			String label= Messages.format(CorrectionMessages.GetterSetterCorrectionSubProcessor_replacewithgetter_description, BasicElementLabels.getJavaCodeString(ASTNodes.asString(context.accessNode)));
+			return createNonNullMethodGetterProposal(label, context.compilationUnit, context.astRewrite, relevance);
+		} else {
+			IJavaElement element= context.variableBinding.getJavaElement();
+			if (element instanceof IField) {
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java:244]: Pattern matching again to be used.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+				return null; //we don't know how to handle those cases.
+			Expression mi= createMethodInvocation(context, method, assignedValue);
+			context.astRewrite.replace(context.accessNode.getParent(), mi, null);
+
+			String label= Messages.format(CorrectionMessages.GetterSetterCorrectionSubProcessor_replacewithsetter_description, BasicElementLabels.getJavaCodeString(ASTNodes.asString(context.accessNode)));
+			T proposal= createMethodSetterProposal(label, context.compilationUnit, context.astRewrite, relevance);
+			return proposal;
+		} else {
+			IJavaElement element= context.variableBinding.getJavaElement();
+			if (element instanceof IField) {
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/GetterSetterCorrectionBaseSubProcessor.java:323]: Pattern matching

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/VarargsWarningsBaseSubProcessor.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+				}
+			} catch (JavaModelException e) {
+				return;
+			}
+		}
+	}
+
+	public void createRemoveSafeVarargsProposals(IInvocationContextCore context, IProblemLocationCore problem, Collection<T> proposals) {
+		ASTNode coveringNode= problem.getCoveringNode(context.getASTRoot());
+		if (!(coveringNode instanceof MethodDeclaration))
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/VarargsWarningsBaseSubProcessor.java:99]: Pattern matching please.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/util/TypeFilter.java [hunk 11]
```diff
@@ -113,7 +113,7 @@ private IEclipsePreferences getDefaultPreferenceStore() {
 
 	private String getPreference(String key, String def) {
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/util/TypeFilter.java:114]: While I don't mind formatting changes in general. This one creates just noise as it doesn't look related to the rest of the PR and no other change is done to the file.

File: org.eclipse.jdt.core.manipulation/proposals/org/eclipse/jdt/internal/ui/text/correction/proposals/RefactoringCorrectionProposalCore.java [hunk 12]
```diff
@@ -67,10 +67,10 @@ public TextChange createTextChange() throws CoreException {
 		Change o = fRefactoring.createChange(new NullProgressMonitor());
 		if(o instanceof TextChange)
 			return (TextChange) o;
-		if( o instanceof CompositeChange) {
+		if (o instanceof CompositeChange) {
```
Comment (akurtakov) [inline org.eclipse.jdt.core.manipulation/proposals/org/eclipse/jdt/internal/ui/text/correction/proposals/RefactoringCorrectionProposalCore.java:70]: Whitespace changes without anything else and it even changes one line to certain formatting but leaves the previous one with different style. Please drop the whole file from this PR as there is nothing but this.

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/java/correction/CUCorrectionProposal.java [hunk 13]
```diff
@@ -253,7 +253,7 @@ private boolean performValidateEdit(ICompilationUnit unit) {
 	 * @throws CoreException if the creation of the text change failed
 	 */
 	protected synchronized TextChange createTextChange() throws CoreException {
-		if( useDelegateToCreateTextChange() ) {
+		if (useDelegateToCreateTextChange()) {
```
Comment (akurtakov) [inline org.eclipse.jdt.ui/ui/org/eclipse/jdt/ui/text/java/correction/CUCorrectionProposal.java:256]: Again a whitespace only change. Please consider submitting a separate PR that does formatting changes where needed.

**Review by akurtakov — COMMENTED** [review]
Formal review from my side without looking in the code details.

**Review by jjohnstn — APPROVED** [review]
